README.md
# encryptsocks

[![npm-version](https://img.shields.io/npm/v/encryptsocks.svg?style=flat-square)](https://www.npmjs.com/package/encryptsocks)
[![Build Status](https://travis-ci.org/oyyd/encryptsocks.svg?branch=master)](https://travis-ci.org/oyyd/encryptsocks)

Encrypt your socks transmission.

* [Why another Nodejs implementation? (with Benchmark)](https://github.com/oyyd/encryptsocks#why-another-nodejs-implementation)
* [CLI](https://github.com/oyyd/encryptsocks#cli)
* [Examples](https://github.com/oyyd/encryptsocks#examples)
* [Config](https://github.com/oyyd/encryptsocks#config)
* [SOCKS5 Username Password Authetication](https://github.com/oyyd/encryptsocks#socks5-username-password-authetication)

## Why another Nodejs implementation?

__Nodejs is a very good choice to achieve both flexibility and good performance in this situation__.

And I have found that many of who are familiar with the [original implementation](https://github.com/shadowsocks/shadowsocks-nodejs) may be curious about the memory usage so that I have finished some simple benchmarks to measure its behavior.

### Benchmark

You can get the benchmark details [here](benchmark/README.md) or even test your own  implementation.

After some simple benchmarks that compare both the node and python implementation, my conclusion is:

1. Node has a different GC strategy but it's, of course, able to keep thousands of connections with a reasonable memory usage. [It's not a bug, it's a conscious time/space trade-off](https://github.com/nodejs/node-v0.x-archive/issues/4525).

2. Each request would cost less time to get responsed (even 50% less time in some situations).

3. Node implementation is less likely to fail requests in high concurrency situation.

And the higher concurrency benchmarks may be meaningless as the bandwidth and network environment would become the actual bottleneck in the real world.

**Do Please** point out my faults if I have missed something or get something wrong.

## Requirement

node >= v4

It's recommended to use node v6 to achieve better performance.

## Installation

```
npm i -g encryptsocks
```

## About the daemon

Encryptsocks use `pm2` as the watcher process from `1.4.0`.

## CLI

Use `localssjs` (local ssjs) to start clients to communicate with applications. The `localssjs` server will also serve a [pac](https://en.wikipedia.org/wiki/PAC) file at `http://127.0.0.1:8090` (by default) for your apps to avoid unnecessary tunnel work.

You may prefer to navigate [clients page](https://shadowsocks.org/en/download/clients.html) and choose clients for your devices instead of using `localssjs`.

Use `serverssjs` (server ssjs) to start your remote server.

Use `localssjs -h` or `serverssjs -h` to show cli options:

```
Proxy options:
  -c config                     path to config file
  -s SERVER_ADDR                server address, default: 127.0.0.1
  -p SERVER_PORT                server port, default: 8083
  -l LOCAL_ADDR                 local binding address, default: 127.0.0.1
  -b LOCAL_PORT                 local port, default: 1080
  -k PASSWORD                   password
  -m METHOD                     encryption method, default: aes-128-cfb
  -t TIMEOUT                    timeout in seconds, default: 600
  --pac_port PAC_PORT           PAC file server port, default: 8090
  --pac_update_gfwlist [URL]    [localssjs] Update the gfwlist
                                for PAC server. You can specify the
                                request URL.
  --level LOG_LEVEL             log level, default: warn
                                example: --level verbose
General options:
  -h, --help                    show this help message and exit
  -d start/stop/restart         daemon mode
```

### Examples

Start clients that bind at `1088` and will connect to `MY.SSSERVER.DOMAIN`:

```
$ localssjs -b 1088 -s MY.SSSERVER.DOMAIN
```

Start daemon:

```
$ localssjs -d start -b 1080
```

Log verbosely:

```
$ serverssjs -d start --level verbose
```

Update GFWList for your .pac file server:

```
$ localssjs --pac_update_gfwlist
```

Update GFWList for your .pac file server from a specific URL (default [url](https://raw.githubusercontent.com/gfwlist/gfwlist/master/gfwlist.txt)):

```
$ localssjs --pac_update_gfwlist http://firefoxfan.cc/gfwlist/gfwlist.txt
```

## Config

```json
{
  "serverAddr": "127.0.0.1",
  "serverPort": 8083,
  "localAddr": "127.0.0.1",
  "localPort": 1080,
  "pacServerPort": 8090,
  "password": "YOUR_PASSWORD_HERE",
  "timeout": 600,
  "method": "aes-128-cfb",

  "level": "warn",
  "localAddrIPv6": "::1",
  "serverAddrIPv6": "::1"
}
```

Specify your config file with `-c` flag:

```
$ serverssjs -c config.json
```

You can change default config in `config.json` file of your global
package.

## SOCKS5 Username Password Authetication

__NOTE:__ This authetication is dangerous when sniffed.

Add `auth` property to your `config.json` and make `forceAuth` `true`.

```json
{
  "auth": {
    "forceAuth": true,
    "usernamePassword": {
      "name": "password"
    }
  }
}
```

## [Optimizing](https://github.com/Long-live-shadowsocks/shadowsocks/wiki/Optimizing-Shadowsocks)

## Encryption methods

* aes-128-cfb
* aes-192-cfb
* aes-256-cfb
* bf-cfb
* camellia-128-cfb
* camellia-192-cfb
* camellia-256-cfb
* cast5-cfb
* des-cfb
* idea-cfb
* rc2-cfb
* rc4
* rc4-md5
* seed-cfb

## Test

```
$ npm test
```

## Contribute

```
$ npm run watch
```

## About the support to UDP relay

I intend to implement UDP relay and I have implement it.
but I can't find an effective way to test this in real world networking.
Please create issues to help us if you know any applications that support
UDP-socks well.

## License

BSD


CHANGELOG.md
# CHANGELOG

## 1.4.2

## 1.4.1@alpha

* Core: Support passing a `proxyOptions` in `pm` instead of creating a config object from `argv`.

## 1.4.0

* Core: Allow to set log path.
* Core: Store list in base64.
* Core: Manage process with `pm2`.

## 1.2.0

* Bug fix: Respect the "serverAddr" option and set default "serverAddr" to "0.0.0.0".

## 1.1.4

* Core: Support http-proxy.

## 1.1.3

* Core: Support SOCKS5 username/password authetication.

## 1.1.2

* Core: Add domain resolving support.

* Bug fix: The option `-c` resolved path incorrectly.

* Bug fix: Windows do not accept kill signals.

## 1.1.1

* Bug fix: `getDstInfo` may return buffers with zero length and throw uncaught error when reading these buffers

## 1.1.0

* Core: Add .pac file server
  * Update rules from gfwlist
  * Support adding user rules

* Core: Seperate log files

* Extra: Add benchmark

## 1.0.4(2016-05-21)

* Bug fix: typo `clientToDst.resumse()`

* Babel: Enable babel loose mode.

## 1.0.3(2016-05-21)

* Change: Do not log timeout warning.


src/cli.js
import { version } from '../package.json';
import * as ssLocal from './ssLocal';
import * as ssServer from './ssServer';
import { updateGFWList as _updateGFWList, GFWLIST_FILE_PATH } from './gfwlistUtils';
import { getConfig, DAEMON_COMMAND } from './config';
import { start, stop, restart } from './pm';

const log = console.log; // eslint-disable-line

function getDaemonType(isServer) {
  return isServer ? 'server' : 'local';
}

function logHelp(invalidOption) {
  log(
// eslint-disable-next-line
`
${(invalidOption ? `${invalidOption}\n` : '')}encryptsocks ${version}
You can supply configurations via either config file or command line arguments.

Proxy options:
  -c CONFIG_FILE                Path to the config file.
  -s SERVER_ADDR                Server address. default: 0.0.0.0
  -p SERVER_PORT                Server port. default: 8083
  -l LOCAL_ADDR                 Local binding address. default: 127.0.0.1
  -b LOCAL_PORT                 Local port. default: 1080
  -k PASSWORD                   Password.
  -m METHOD                     Encryption method. default: aes-128-cfb
  -t TIMEOUT                    Timeout in seconds. default: 600
  --pac_port PAC_PORT           PAC file server port. default: 8090
  --pac_update_gfwlist [URL]    [localssjs] Update the gfwlist
                                for PAC server. You can specify the
                                request URL.
  --log_path LOG_PATH           The directory path to log. Won't if not set.
  --level LOG_LEVEL             Log level. default: warn
                                example: --level verbose
General options:
  -h, --help                    Show this help message and exit.
  -d start/stop/restart         Run as a daemon.
`
  );
}

function updateGFWList(flag) {
  log('Updating gfwlist...');

  const next = (err) => {
    if (err) {
      throw err;
    } else {
      log(`Updating finished. You can checkout the file here: ${GFWLIST_FILE_PATH}`);
    }
  };

  if (typeof flag === 'string') {
    _updateGFWList(flag, next);
  } else {
    _updateGFWList(next);
  }
}

function runDaemon(isServer, cmd) {
  const type = getDaemonType(isServer);

  switch (cmd) {
    case DAEMON_COMMAND.start: {
      start(type);
      return;
    }
    case DAEMON_COMMAND.stop: {
      stop(type);
      return;
    }
    case DAEMON_COMMAND.restart: {
      restart(type);
      break;
    }
    default:
  }
}

function runSingle(isServer, proxyOptions) {
  const willLogToConsole = true;
  return isServer ? ssServer.startServer(proxyOptions, willLogToConsole)
    : ssLocal.startServer(proxyOptions, willLogToConsole);
}

export default function client(isServer) {
  const argv = process.argv.slice(2);

  getConfig(argv, (err, config) => {
    if (err) {
      throw err;
    }

    const { generalOptions, proxyOptions, invalidOption } = config;

    if (generalOptions.help || invalidOption) {
      logHelp(invalidOption);
    } else if (generalOptions.pacUpdateGFWList) {
      updateGFWList(generalOptions.pacUpdateGFWList);
    } else if (generalOptions.daemon) {
      runDaemon(isServer, generalOptions.daemon);
    } else {
      runSingle(isServer, proxyOptions);
    }
  });
}
